### ISSUE TITLE
Custom AuthenticationEntryPoint not set on BasicAuthenticationFilter

### ISSUE THREAD

**Issue by jigishpa** [spring-projects/spring-security-oauth#501]
I am trying to use a custom AuthenticationEntryPoint for Authorization Server / provider. I tried to set it using the Configurer as follows:

```
@Configuration
@EnableAuthorizationServer
public class AuthorizationServerConfig extends AuthorizationServerConfigurerAdapter{
    ...
    @Autowired
    MyBasicAuthenticationEntryPoint myBasicAuthenticationEntryPoint;

    @Override
    public void configure(AuthorizationServerSecurityConfigurer security) throws Exception {
        security.authenticationEntryPoint(myBasicAuthenticationEntryPoint);
    }
    ...
}
```

I tried stepping through the code. If I use invalid credentials to obtain a token, I notice that a DelegatingAuthenticationEntryPoint is invoked which in turn invokes an instance of default BasicAuthenticationEntryPoint (instead of my version of the same).

[Sample code in github](https://github.com/jigishpa/spring-security-oauth). Relevant AuthorizationServer [configuration here](https://github.com/jigishpa/spring-security-oauth/blob/master/authorization-server-basic/src/main/java/com/senyume/oauth2/authorization/AuthorizationServerConfig.java)

I also noticed that there's a somewhat [similar issue open in github](https://github.com/spring-projects/spring-security-oauth/issues/112)

The type of entry point doesn't matter. I tried something similar with a class that extends OAuth2AuthenticationEntryPoint but same behavior.

**Comment by dsyer** [comment]
Which grant type are you using, and which credentials are wrong (the client or the user)?

Issue #112 is unrelated (it's about OAuth1).

**Comment by jigishpa** [comment]
I am using client credentials grant. In the sample code I posted, valid credentials for obtaining a bearer token are "my-client-with-secret:secret". When I intentionally pass incorrect credentials, the authorization server doesn't seem to trigger my custom basic authentication endpoint.

My goal is to customize the format of the error messages. I am able to do that with the Resource Server by doing something similar.

**Comment by EmilIfrim** [comment]
I have the same issue. The AuthorizationServer uses BasicAuthenticationFilter and I don't know how to set my custom AuthenticationEntryPoint. I have almost the same setup as the creator of this issue. Using this setup only works when ClientDetails are sent as parameters and not as Authorization header. What is the right way to solve this?

**Comment by dai0304** [comment]
+1
Is there any workaround?

**Comment by ChanggyunCho** [comment]
Hello

I've using spring-security-oauth2 2.2.0.RELEASE.
The custom AuthenticationEntryPoint is also not available when allowFormAuthenticationForClients is set in the AuthorizationServer configuration.

```java
    @Override
    public void configure(AuthorizationServerSecurityConfigurer security) throws Exception {
        //authEntryPoint is my custom AuthenticationEntryPoint
        security.allowFormAuthenticationForClients().authenticationEntryPoint(authEntryPoint);
    }
```

With this flag I can avoid BasicAuthenticationFilter issue for bad credential, but my custom authentication entry point still not be processed due to the ClientCredentialsTokenEndpointFilter in AuthorizationServerSecurityConfigurer

```java
	@Override
	public void configure(HttpSecurity http) throws Exception {
                ...
		if (allowFormAuthenticationForClients) {
			clientCredentialsTokenEndpointFilter(http);
		}
                ...
	}
```

That function creates ClientCredentialsTokenEndpointFilter with hard-coded Oauth2AuthenticationEntryPoint.
So I cannot get my customized error message for wrong client_id or secret with client_credentials.

```java
	private ClientCredentialsTokenEndpointFilter clientCredentialsTokenEndpointFilter(HttpSecurity http) {
                ...
		OAuth2AuthenticationEntryPoint authenticationEntryPoint = new OAuth2AuthenticationEntryPoint();
		authenticationEntryPoint.setTypeName("Form");
		authenticationEntryPoint.setRealmName(realm);
		clientCredentialsTokenEndpointFilter.setAuthenticationEntryPoint(authenticationEntryPoint);
                ...
	}
```

Is there any workaround or plan to fix this issue?

Thanks and regards,
Changgyun

**Comment by brankoiliccc** [comment]
+1
Is there a solution for this issue ?

**Comment by mikelhamer** [comment]
+1
Please fix

**Comment by meose** [comment]
+1
I faced with the same problem.
No solution yet?

**Comment by hectorarguello** [comment]
+1
I have three days trying to implement custom response with exactly the same scenario @ChanggyunCho describes.
Please provide a solution.
Regards.

**Comment by eugeniosant** [comment]
+1
I have the same problem! Is there a solution?

**Comment by Van-Dame** [comment]
It's been more than an year. Any solutions to this? Why would the AuthenticationEntryPoint be hard-coded in the AuthorizationServerSecurityConfigurer.clientCredentialsTokenEndpointFilter?

**Comment by jgrandja** [comment]
If an `AuthenticationEntryPoint` is supplied via `AuthorizationServerSecurityConfigurer.authenticationEntryPoint()` than it's associated with the `ExceptionTranslationFilter`. However, it also needs to be associated with the `BasicAuthenticationFilter`, as this is the filter that handles client authentication when the client attempts to authenticate using HTTP Basic.

As a temporary workaround, here is a [sample](https://github.com/jgrandja/spring-security-oauth-sample/tree/gh-501-custom-authentrypoint) that demonstrates how you can set a custom `AuthenticationEntryPoint` with the `BasicAuthenticationFilter` in order to have the ability to customize the response for a failed client authentication.

Take a look at the `AuthorizationServerConfigurerAdapter` and specifically this [line](https://github.com/jgrandja/spring-security-oauth-sample/blob/gh-501-custom-authentrypoint/auth-server/src/main/java/org/springframework/security/oauth/samples/config/AuthorizationServerConfig.java#L55) for the workaround.

**Comment by jgrandja** [comment]
Apologies for the delay here but this is now in master.

**Comment by vicykie** [comment]
+1

I have the same problem! Is there a solution?

**Comment by xuebingchen** [comment]
@ChanggyunCho 
My problem is the same as yours, can you tall me how did you solve it?

**Comment by nicejing** [comment]
> @ChanggyunCho
> My problem is the same as yours, can you tall me how did you solve it?

+1 
How to do ?

**Comment by nicejing** [comment]
@jgrandja Hello, I still have the same problem! Which version has solved? My dependency pom id 
```
<dependency>
          <groupId>org.springframework.cloud</groupId>
           <artifactId>spring-cloud-starter-oauth2</artifactId>
           <version> 2.1.3.RELEASE</version>
 </dependency>
```
please let me know, thanks a lot.

**Comment by ningca** [comment]
hi, I am having the same problem with ClientCredentialsTokenEndpointFilter, since the AuthenticationEntryPoint is hardcoded in the code AuthorizationServerSecurityConfigurer.java

	private ClientCredentialsTokenEndpointFilter clientCredentialsTokenEndpointFilter(HttpSecurity http) {
                ...
		**OAuth2AuthenticationEntryPoint authenticationEntryPoint = new OAuth2AuthenticationEntryPoint();**

Please fix it, I am using spring-security-oauth2-2.3.6.RELEASE.jar, thanks.

**Comment by msampietro** [comment]
**WORKAROUND use carefully**

Create a @Configuration class after SecurityFilterChain has been created and override authenticationEntryPoint of ClientCredentialsTokenEndpointFilter based on your chain order.

Example:

`@Configuration
public class ClientCredentialsTokenEndpointFilterWorkaround {

    private final Filter filterChain;

    @Autowired
    public ClientCredentialsTokenEndpointFilterWorkaround(Filter springSecurityFilterChain) {
        this.filterChain = springSecurityFilterChain;
        workaround();
    }

    private void workaround() {
        FilterChainProxy filterChainProxy = (FilterChainProxy) filterChain;
        SecurityFilterChain securityFilterChain = filterChainProxy.getFilterChains().get(4);
        Filter filter = securityFilterChain.getFilters().get(4);
        if (filter instanceof ClientCredentialsTokenEndpointFilter) {
            ClientCredentialsTokenEndpointFilter clientCredentialsTokenEndpointFilter = (ClientCredentialsTokenEndpointFilter) filter;
            clientCredentialsTokenEndpointFilter.setAuthenticationEntryPoint(new CustomOAuth2AuthenticationEntryPoint());
        }
    }

}`
